### PR TITLE
*: add show-ddl-locks support for optimistic mode

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

add `show-ddl-locks` support for optimistic mode

### What is changed and how it works?

- construct `synced` and `unsynced` based on source, schema and table name.
- add a `mode` field for the lock.

```
» show-ddl-locks
{
    "result": true,
    "msg": "",
    "locks": [
        {
            "ID": "sequence_sharding_optimistic-`sharding_target_opt`.`t_target`",
            "task": "sequence_sharding_optimistic",
            "mode": "optimistic",
            "owner": "",
            "DDLs": [
            ],
            "synced": [
                "mysql-replica-01-`sharding_seq_opt`.`t2`",
                "mysql-replica-02-`sharding_seq_opt`.`t2`",
                "mysql-replica-02-`sharding_seq_opt`.`t4`"
            ],
            "unsynced": [
                "mysql-replica-01-`sharding_seq_opt`.`t1`",
                "mysql-replica-02-`sharding_seq_opt`.`t3`"
            ]
        }
    ]
}
```

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@WangXiangUSTC @lichunzhu PTAL

File: dm/master/server.go [hunk 1]
```diff
@@ -684,6 +683,9 @@ func (s *Server) ShowDDLLocks(ctx context.Context, req *pb.ShowDDLLocksRequest)
 		resp.Locks = append(resp.Locks, l)
 	}
 
+	// show optimistic locks.
+	resp.Locks = append(resp.Locks, s.optimist.ShowLocks(req.Task, req.Sources)...)
```
Comment (lichunzhu) [inline dm/master/server.go:687]: Should we use some field to differ optimist lock from pessimist ones? Since their definitions are different.

Comment (csuzhangxc) [reply dm/master/server.go:687]: how about adding a field named `mode`?

Comment (lichunzhu) [reply dm/master/server.go:687]: I think it's OK

Comment (csuzhangxc) [reply dm/master/server.go:653]: addressed in https://github.com/pingcap/dm/pull/580/commits/a02a9cc269f8647a8cd9b6d30cae116eb1a34df8.

File: dm/master/shardddl/optimist.go [hunk 2]
```diff
@@ -104,6 +109,51 @@ func (o *Optimist) Locks() map[string]*optimism.Lock {
 	return o.lk.Locks()
 }
 
+// ShowLocks is used by `show-ddl-locks` command.
+func (o *Optimist) ShowLocks(task string, sources []string) []*pb.DDLLock {
```
Comment (WangXiangUSTC) [inline dm/master/shardddl/optimist.go:113]: do we need to add read lock?

File: dm/master/shardddl/pessimist.go [hunk 3]
```diff
@@ -217,6 +219,47 @@ func (p *Pessimist) Locks() map[string]*pessimism.Lock {
 	return p.lk.Locks()
 }
 
+// ShowLocks is used by `show-ddl-locks` command.
+func (p *Pessimist) ShowLocks(task string, sources []string) []*pb.DDLLock {
```
Comment (WangXiangUSTC) [inline dm/master/shardddl/pessimist.go:223]: ditto

File: dm/master/shardddl/optimist.go (hunk 2 above)
Comment (csuzhangxc) [reply dm/master/shardddl/optimist.go:113]: use read lock to protect what?

Comment (WangXiangUSTC) [reply dm/master/shardddl/optimist.go:113]: ok, a mistake, the `o.lk.locks` returns a copy, don't need to protect

**Pull request merged**